### ISSUE TITLE
Testing current 000 bootable build on Hatari

### ISSUE THREAD

**Issue by skaftetryne** [freemint/freemint#377]
1. There is a mint000h.prg in SYSDIR, but it does not appear to be loaded as hostfs is not working. If I rename this to mint000.prg XaAES will crash during boot. How is this supposed to work? 

![image](https://github.com/user-attachments/assets/a81d9876-990b-48d7-89e3-3fbd4d13961e)

There is no XaAES log file, neither in C:\ nor in SYSDIR/xaaes/. I have enabled "logfile = $(BOOTDRIVE):\xaaes.log" in xaaes.cnf.

2. I get this alert when XaAES starts:

![image](https://github.com/user-attachments/assets/99bec48a-a112-47b1-b358-3376dc946b17)

I believe this is from NVDI itself and not via the alert pipe. The function in question is vex_wheelv, which is not present.

3. When doing a shutdown from the task manager I get this:

![image](https://github.com/user-attachments/assets/073c167d-b31c-4f53-9081-18c86fda77b8)

Selecting "Kill'em all" results in yet another appearance of this:

![image](https://github.com/user-attachments/assets/27d7ad76-71c3-4c87-bdf7-001b4c6133b1)

The alert in not responsive so shutdown stops here. Disabling moose_w.adi fixes (2) and (3).

**Comment by mikrosk** [comment]
Oh man. And mint030h.prg is free of this issues?

From what I can see, the 000h version is compiled with the same flags but it contains additional `NO` flags (like ramdisk) to reduce it file size. Maybe some of them are not playing nicely with Hatari.

However if renaming definitely shouldn't have any effect.

**Comment by th-otto** [comment]
The error message for `vex_wheelv` can be disabled using NVDICONF.CPX.

Fixing that in mint seems difficult. The function is only present in fVDI, however you can have both fVDI and NVDI running, to use NVDI for vector fonts, or its printer drivers. And you can also configure fVDI to install a NVDI cookie, since some programs require this. So how to check that?

**Comment by skaftetryne** [comment]
> The error message for `vex_wheelv` can be disabled using NVDICONF.CPX.

2.51 does not have the CPX, only NVDICONF.ACC.  And it will only disable the error message at shutdown, since the settings are applied by the ACC and the first occurence is before any accessories has been loaded.

> Fixing that in mint seems difficult. The function is only present in fVDI,

It is also present in MilanTOS, it was in fact first introduced in MilanTOS. You can check for the presence of either fVDI or a Milan before attempting this call, but what if someone creates a TSR that adds this call to other VDI's? The alert from NVDI 2.51 is not really the problem I think, AFAICS later versions does not show this alert. The real problem is that AESSYS does not respond correctly to a shutdown, and I wonder if that is because it's trying to remove itself from the vex_wheel vector. So a fix could be to not do this when the attempt to hook into this vector failed.

**Comment by skaftetryne** [comment]
> Oh man. And mint030h.prg is free of this issues?

I have not tested, I'm setting up a disk image for a plain ST. But I could test this later.

> However renaming definitely shouldn't have any effect.

So mintloader will detect the presence of Hatari and load mint000h.prg automatically?

**Comment by mikrosk** [comment]
> So mintloader will detect the presence of Hatari and load mint000h.prg automatically?

Surely not. That's the problem with Hatari, they don't want to be discovered as emulator. :)

> If I rename this to mint000.prg XaAES will crash during boot. How is this supposed to work?

Aaaah, now I get it. You had mint000h.prg in $SYSDIR, assuming that it will load, using the old gemdos filesystem. However what happen is that mint000.prg was loaded, therefore it's clear why it didn't work.

Then you renamed 000h to 000 and it crashed because it was loaded this time. Therefore there's something really wrong with the 000 Hatari build.

What I don't understand is how you loaded XaAES if gemdos filesystem was not active and therefore Hatari couldn't emulate it's host fs?

**Comment by skaftetryne** [comment]
> Then you renamed 000h to 000 and it crashed because it was loaded this time. Therefore there's something really wrong with the 000 Hatari build.

Exactly.

> What I don't understand is how you loaded XaAES if gemdos filesystem was not active and therefore Hatari couldn't emulate it's host fs?

I'm booting from a disk image, not from hostfs. So I have partitions C: to F: on the disk image, and - with MiNT 1.12 - a hostfs drive on I:

**Comment by th-otto** [comment]
>The real problem is that AESSYS does not respond correctly to a shutdown, and I wonder if that is because it's trying to > 
remove itself from the vex_wheel vector. So a fix could be to not do this when the attempt to hook into this vector failed.

But the problem is that XaAES does not simply get an error code when trying to deinstall the vex_wheelv vector, instead it is blocked by the alert from NVDI.

What could be done though, is to temporarily change the nvdi configuration during the call, to prevent the alert.

**Comment by skaftetryne** [comment]
Even with the alert supressed AESSYS still does not respond to a shutdown properly.

> But the problem is that XaAES does not simply get an error code 

No, but it does save the address to the old handler in the variable whose address is passed to it in this call. So one can set this variable to a known value before calling vex_wheel, and see if it has been changed when the call returns.

**Comment by th-otto** [comment]
> There is no XaAES log file, neither in C:\ nor in SYSDIR/xaaes/. I have enabled "logfile = $(BOOTDRIVE):\xaaes.log" in xaaes.cnf.

Just noticed this.  Did you really use '('? The new support for variables needs curly braces '{'

**Comment by th-otto** [comment]
> No, but it does save the address to the old handler in the variable whose address is passed to it in this call.

If the function is not supported, the call does nothing (except reporting the error maybe). Especially is does not set the old vector to anything at all.

**Comment by skaftetryne** [comment]
Excatly. You will know that the call has failed if the variable you passed the adress of to the call is not changed when the call returns.

![image](https://github.com/user-attachments/assets/cd11b260-ad8d-47d8-81f5-0d4cef93a5f9)

**Comment by skaftetryne** [comment]
> Just noticed this. Did you really use '('? The new support for variables needs curly braces '{'

No, a typo from me. It's using curly braces, I just uncommented the setting from the default xaaes.cnf.

**Comment by th-otto** [comment]
> Excatly. You will know that the call has failed if the variable you passed the adress of to the call is not changed when the call returns.

But that's not how the binding works. The return address is part of the control structure, which is a local variable in gemlib. If the VDI call does not write anything there, gemlib will return random garbage from the stack (which i just fixed).
Also just checking the address would not have helped, you would still get an error dialog from NVDI. This too, should be fixed now.

**Comment by mikrosk** [comment]
Can you post more messages? I.e. the ones since the launch of xaloader.prg?

**Comment by skaftetryne** [comment]
Boot-log (btw. I had to search for this, finally found it in the machine (ST) subfolder):

`Calibrating delay loop ... 3.12 BogoMIPS
`
`Initializing built-in domain ops ... done!
`Loading external modules ...
`
`p xconout2 output catcher version 0.3 q
`½ Jan  1 1995 by TeSche <itschere@techfak.uni-bielefeld.de>
`½ 2000-2010 by Frank Naumann.
`
`p Ext2 filesystem driver version 0.63 q
`½ 1998, 1999 by Axel Kaiser.
`½ 2000-2010 by Frank Naumann.
`½ 2013 by Alan Hourihane.
`
`p Minix file system driver for MiNT. q
`Version 0.75
`Copyright 1991,1992,1993,1994,1995 S.N.Henson
`pThis is a modified binary version for FreeMiNT 1.15.q
`½ Sep  1 2024 by Frank Naumann.
`
`
`Starting up the update daemon ... done!
`Starting up the idle process (pid 0) ... done!
`
`Reading `u:/c/mint/1-19-7eb/mint.cnf' ... 65 bytes done.
`
`
`Reading `u:/c/mint/1-19-7eb/mint_def.cnf' ... 15150 bytes done.
`
`GEM 'u:/c/mint/1-19-7eb/xaaes/xaloader.prg' ''
`Setup complete, now booting the system...
`
`init is gem and init_prg = u:/c/mint/1-19-7eb/xaaes/xaloader.prg
`

Debug log:
`pid   0 (MiNT): ext2 (main.c): init
`pid   0 (MiNT): ext2 (main.c): running in native UTC mode!
`pid   0 (MiNT): ext2 (main.c): loaded and ready (k = 674B0) -> D1E68.
`pid   0 (MiNT): Minix-FS: running in native UTC mode!
`pid   0 (MiNT): Minix-FS: loaded and ready (k = 674B0).
`pid   0 (MiNT): parse_cnf: can't open u:/c/mint/1-19-7eb/st/mint.cnf
`pid   1 (xaloader): run_km(xaaes.km) ok (bp 0x14E360)!
`pid   1 (xaloader): run_km: run=0x14E460
`pid   4 (AESSYS): XaAES memory leaks: 145 bytes
`pid   0 (MiNT): init:sys_s_hutdown:COLD
`pid   0 (MiNT): p->name=NVDI:247A1C pgrp=0
`pid   0 (MiNT): p->name=update:EB01C pgrp=0
`pid   0 (MiNT): p->name=MiNT:92FF0 pgrp=0
`pid   0 (MiNT): Close open files ...
`pid   0 (MiNT): done
`pid   0 (MiNT): Syncing file systems ...
`pid   0 (MiNT): done
`pid   0 (MiNT): Unmounting C: ...
`pid   0 (MiNT): Syncing file systems ...
`pid   0 (MiNT): done
`
